File: tests/test_gateway_routing.py
import unittest
from unittest.mock import AsyncMock, patch, MagicMock
import asyncio

class TestTelegramGateway(unittest.IsolatedAsyncioTestCase):
    """TestTelegramGateway class."""
    def setUp(self):
        self.whitelist = [123456789] # Mock authorized User ID
        self.unauthorized_user = 987654321

    @patch('zena_mode.gateway_telegram.backend') # Mocking the global backend
    async def test_authorized_user_routing(self, mock_backend):
        """Verify that an authorized user's message is sent to the LLM."""
        from zena_mode.gateway_telegram import handle_telegram_message
        
        # Patch the whitelist inside the module for the test
        with patch('zena_mode.gateway_telegram.ALLOWED_USER_IDS', [123456789]):
            # Mock backend streaming response
            mock_gen = MagicMock()
            mock_gen.__aiter__.return_value = ["Hello ", "from ZenAI!"]
            mock_backend.send_message_async.return_value = mock_gen
            mock_backend.__aenter__.return_value = mock_backend
            
            # Mock Telegram Message/Update
            update = MagicMock()
            update.effective_user.id = 123456789
            update.message.text = "Tell me about the swarm."
            update.message.reply_text = AsyncMock()
            
            context = MagicMock()
            context.bot.send_message = AsyncMock()
            context.bot.edit_message_text = AsyncMock()

            await handle_telegram_message(update, context)
            
            # Check if backend was called
            mock_backend.send_message_async.assert_called_with("Tell me about the swarm.")

    async def test_unauthorized_user_rejection(self):
        """Ensure that users not in the whitelist are blocked."""
        from zena_mode.gateway_telegram import handle_telegram_message
        
        with patch('zena_mode.gateway_telegram.ALLOWED_USER_IDS', [123456789]):
            update = MagicMock()
            update.effective_user.id = 987654321 # Unauthorized
            update.message.text = "Hack me!"
            update.message.reply_text = AsyncMock()

            context = MagicMock()
            context.bot.send_message = AsyncMock()

            await handle_telegram_message(update, context)
            
            # Should have sent a rejection message
            args, kwargs = update.message.reply_text.call_args
            self.assertIn("Unauthorized", args[0])

if __name__ == "__main__":
    unittest.main()

File: rust_output/src/test_gateway_routing.rs
use anyhow::{Result, Context};
use tokio;

/// TestTelegramGateway class.
#[derive(Debug, Clone)]
pub struct TestTelegramGateway {
}

impl TestTelegramGateway {
    pub fn setUp(&mut self) -> () {
        self.whitelist = vec![123456789];
        self.unauthorized_user = 987654321;
    }
    /// Verify that an authorized user's message is sent to the LLM.
    pub async fn test_authorized_user_routing(&self, mock_backend: String) -> () {
        // Verify that an authorized user's message is sent to the LLM.
        // TODO: from zena_mode.gateway_telegram import handle_telegram_message
        /* let _ctx = mock::/* mock::patch(...) */ — use mockall crate */;
        {
            let mut mock_gen = MagicMock();
            mock_gen.__aiter__.return_value = vec!["Hello ".to_string(), "from ZenAI!".to_string()];
            mock_backend::send_message_async.return_value = mock_gen;
            mock_backend::__aenter__.return_value = mock_backend;
            let mut update = MagicMock();
            update.effective_user.id = 123456789;
            update.message.text = "Tell me about the swarm.".to_string();
            update.message.reply_text = AsyncMock();
            let mut context = MagicMock();
            context.bot.send_message = AsyncMock();
            context.bot.edit_message_text = AsyncMock();
            handle_telegram_message(update, context).await;
            mock_backend::send_message_async.assert_called_with("Tell me about the swarm.".to_string());
        }
    }
    /// Ensure that users not in the whitelist are blocked.
    pub async fn test_unauthorized_user_rejection(&mut self) -> () {
        // Ensure that users not in the whitelist are blocked.
        // TODO: from zena_mode.gateway_telegram import handle_telegram_message
        /* let _ctx = mock::/* mock::patch(...) */ — use mockall crate */;
        {
            let mut update = MagicMock();
            update.effective_user.id = 987654321;
            update.message.text = "Hack me!".to_string();
            update.message.reply_text = AsyncMock();
            let mut context = MagicMock();
            context.bot.send_message = AsyncMock();
            handle_telegram_message(update, context).await;
            let (mut args, mut kwargs) = update.message.reply_text.call_args;
            assert!(args[0].contains("Unauthorized".to_string()));
        }
    }
}
